### PR TITLE
Attendee Tracking: Created AttendeeSelectionPage

### PR DESCRIPTION
**Pull request by zfurtak**
### Fixed Issues

$ https://github.com/Expensify/App/issues/47171
$ https://github.com/Expensify/App/issues/47172
PROPOSAL:

### Tests

1. Click `+` -> Submit expense
2. Choose workspace, amount
3. On expense summary screen, expand options
4. Attendees menu item should be seen with displayName of the current user(who are creating the expanse)
5. Click on the attendees
6. Attendees selection page should appear with this user user selected
7. Also on the screen should be seen sections `Recents` and `Contacts` without workspaces in them
8. You should be able to choose someone, then they should appear in the top section
9. Click Save -> Attendees button should have now users who were chosen
- [x] Verify that no errors appear in the JS console

### QA Steps

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

### CODE REVIEW

File: src/pages/iou/request/step/IOURequestStepAttendees.tsx
Comment (blazejkustra) [inline src/pages/iou/request/step/IOURequestStepAttendees.tsx:19]: 👀

File: src/libs/actions/IOU.ts
Comment (blazejkustra) [inline src/libs/actions/IOU.ts:3001]: Would that work?
```suggestion
    attendees: Attendee[],
```

File: src/libs/OptionsListUtils.ts
Comment (blazejkustra) [inline src/libs/OptionsListUtils.ts:2509]: Is this WIP?

File: src/libs/OptionsListUtils.ts
Comment (blazejkustra) [inline src/libs/OptionsListUtils.ts:1671]: Make sure it doesn't break filtering logic for participants screen and others (by looking at the usage)

File: src/components/MoneyRequestConfirmationListFooter.tsx
Comment (blazejkustra) [inline src/components/MoneyRequestConfirmationListFooter.tsx:514]: Using `\u00B7` was intentional?

File: src/pages/iou/request/MoneyRequestAttendeeSelector.tsx
Comment (blazejkustra) [inline src/pages/iou/request/MoneyRequestAttendeeSelector.tsx:105]: Let's move it to getAttendeesOptions as discussed in person

**Review by blazejkustra — CHANGES_REQUESTED** [review]

File: src/libs/OptionsListUtils.ts
Comment (zfurtak) [reply src/libs/OptionsListUtils.ts]: yes, I changed it in the latest commit

File: src/components/MoneyRequestConfirmationListFooter.tsx
Comment (zfurtak) [reply src/components/MoneyRequestConfirmationListFooter.tsx:514]: Yes, it is unicode for middle dot symbol

File: src/pages/iou/request/MoneyRequestAttendeeSelector.tsx
Comment (zfurtak) [reply src/pages/iou/request/MoneyRequestAttendeeSelector.tsx]: Done in the latest commit!

**Comment by lakchote** [comment]
@ZhenjaHorbach please review it too, this is an important PR that needs to be merged by Monday (hopefully).

**Comment by lakchote** [comment]
> > @ZhenjaHorbach please review it too, this is an important PR that needs to be merged by Monday (hopefully).
> 
> Got it ! In that case I'll check today

Thank you very much!

**Comment by dylanexpensify** [comment]
@ZhenjaHorbach did you still need access to the docs?

**Comment by lakchote** [comment]
> > @ZhenjaHorbach did you still need access to the docs?
> 
> 
> 
> Yes, please !
> 
> My email: [horbachyauhenipetrovich@gmail.com](mailto:horbachyauhenipetrovich@gmail.com)

I've added you! Let me know if you need anything else.

File: src/pages/iou/request/MoneyRequestAttendeeSelector.tsx
Comment (zfurtak) [reply src/pages/iou/request/MoneyRequestAttendeeSelector.tsx:36]: Good catch!

File: src/pages/iou/request/MoneyRequestAttendeeSelector.tsx
Comment (zfurtak) [reply src/pages/iou/request/MoneyRequestAttendeeSelector.tsx:196]: Yeah I guess the types and the function name itself are enough

File: src/pages/iou/request/MoneyRequestAttendeeSelector.tsx
Comment (zfurtak) [reply src/pages/iou/request/MoneyRequestAttendeeSelector.tsx:217]: Sure, I will change that but also I need to disable eslint for this line, as we have rule `??` over `||`

**Comment by zfurtak** [comment]
hi @ZhenjaHorbach, thank you for the review! :)
About the previous message, I fixed the first bug, however I'm not sure about the second one as you don't exit the creation of expense, so the transanction-draft object is not cleaned 🤔

File: src/libs/OptionsListUtils.ts
Comment (blazejkustra) [inline src/libs/OptionsListUtils.ts:1995]: It seems this `map` does nothing, why do we need an assertion here?

File: src/libs/OptionsListUtils.ts
Comment (blazejkustra) [inline src/libs/OptionsListUtils.ts:1930]: Passing `includeRecentReports === true` and `recentAttendees` will completely ignore `recentAttendees` argument currently, I think we should move `else if (recentAttendees && recentAttendees?.length > 0) {` to here.

File: src/libs/OptionsListUtils.ts
Comment (blazejkustra) [inline src/libs/OptionsListUtils.ts:1993]: Should we use this logic here as well?
```ts
optionsToExclude.push({login: reportOption.login});
```

**Review by blazejkustra — CHANGES_REQUESTED** [review]

**Comment by filip-solecki** [comment]
> And I found another bugs
> 
> 1. After we unselect submitter of the Request, we will no longer be able to find it in the sections
> 
> But I'm not sure how best to fix this Maybe we need to create a new section 🤔 I think we need help from designers on where to best display a submitter of the Request
> 
> @Expensify/design Could you please check this case and help us how to fix it better?
> 
> 2. In Recents we can find a workspace and after we select the workspace, the list will break
> 
>  2024-09-02.09.16.31.mov

Regarding pt 2 I am wondering if this shouldn't be handled by BE and if Workspaces shpuldn't be excluded from the recents list? Otherwise we have to filter received data somehow
cc @Julesssss WDYT?

File: src/libs/OptionsListUtils.ts
Comment (filip-solecki) [reply src/libs/OptionsListUtils.ts:1995]: Yes, map is useless, assertion is needed because `reportID` is required in OptionData and is optional in Attendee type

**Comment by Julesssss** [comment]
> Regarding pt 2 I am wondering if this shouldn't be handled by BE and if Workspaces shpuldn't be excluded from the recents list?

We'll need to find a way to exclude workspaces here, as attendees can only ever be people.

**Comment by Julesssss** [comment]
I shared the above questions [here](https://github.com/Expensify/App/issues/47171#issuecomment-2324666424) with the design team

File: src/libs/OptionsListUtils.ts
Comment (blazejkustra) [reply src/libs/OptionsListUtils.ts:1995]: How about this?
```suggestion
        recentReportOptions = recentAttendees as ReportUtils.OptionData[];
```

File: src/libs/OptionsListUtils.ts
Comment (blazejkustra) [reply src/libs/OptionsListUtils.ts:1993]: Oh sorry, it's there in `forEach` above

File: src/libs/OptionsListUtils.ts
Comment (filip-solecki) [reply src/libs/OptionsListUtils.ts:1995]: Yes, it works, I thought I've pushed it

**Comment by Julesssss** [comment]
Hey @filip-solecki, as Jon mentioned here, after unselection let's show users their own profile item row always at the top of recent.

Also, FYI I'm going to switch the base branch to a feature branch (not sure if non-org members have the ability to do that)

**Comment by filip-solecki** [comment]
> Hey @filip-solecki, as Jon mentioned [here](https://github.com/Expensify/App/issues/47171#issuecomment-2325391791), after unselection let's show users their own profile item row always at the top of recent.
> 
> Also, FYI I'm going to switch the base branch to a feature branch (not sure if non-org members have the ability to do that)

Done, when creator is unselected it goes to Recents section.

File: src/libs/OptionsListUtils.ts
Comment (filip-solecki) [reply src/libs/OptionsListUtils.ts:1930]: `includeRecentReports` is false now, it should work fine when BE will be ready

**Comment by filip-solecki** [comment]
Ok, thanks I'll work on them

**Comment by filip-solecki** [comment]
Both fixed, please check again @ZhenjaHorbach

File: src/pages/iou/request/MoneyRequestAttendeeSelector.tsx
Comment (dubielzyk-expensify) [reply src/pages/iou/request/MoneyRequestAttendeeSelector.tsx:140]: Yes we should count it as part of the 5. I think the reason for the list being 5 is to ensure it doesn't go super long and that it's easy to see the rest of the contact underneath.

File: src/libs/OptionsListUtils.ts
Comment (filip-solecki) [reply src/libs/OptionsListUtils.ts:2203]: Yes, this is because `nvp_expensify_recentAttendees` is not ready yet, right @Julesssss ?

File: src/pages/iou/request/MoneyRequestAttendeeSelector.tsx
Comment (filip-solecki) [reply src/pages/iou/request/MoneyRequestAttendeeSelector.tsx:140]: Ok

Comment (filip-solecki) [reply src/pages/iou/request/MoneyRequestAttendeeSelector.tsx]: Done, max 5 recents are displaying and the rest goes to the top of contacts section

**Comment by filip-solecki** [comment]
@ZhenjaHorbach again all yours 😄

**Comment by filip-solecki** [comment]
Looks like a bug, let me check

**Comment by filip-solecki** [comment]
@ZhenjaHorbach can you upload a video?

File: src/libs/OptionsListUtils.ts
Comment (zfurtak) [reply src/libs/OptionsListUtils.ts:1930]: I did it on purpose this way cause the cases where we use `includeRecentReports` and `recentAttendees` are mutually exclusive 🤓
Also `includeRecentReports` value is `false` by default, so there's no need to explicitly set it

Comment (blazejkustra) [reply src/libs/OptionsListUtils.ts:1932]: Ultimately I won't block the PR for this, so feel free to do whatever you think makes more sense 😄

**Review by blazejkustra — CHANGES_REQUESTED** [review]
The diff just got hugee (we need to merge the newest main on the feature branch)

Also @zfurtak / @filip-solecki let's add test steps to the description 👀

**Comment by Julesssss** [comment]
Thanks all, this is looking good. As we haven't yet made the backend changes available to you I think we are in a good place to merge to the feature branch. Then we can create bugs for the remaining issues.

Once main is merged I'm going to add a C+ reviewer to try to locate other bugs that may have been mixed.

**Review by ZhenjaHorbach — APPROVED** [review]

**Comment by zfurtak** [comment]
@ZhenjaHorbach don't worry! it was not very elegant but for sure quick fix 😊

> And when I open screen Recents list is empty, but when I unselect submitter of the Request
> I have full Recents list
About this bug, I couldn't reproduce it, did you do anything special? 😅

> When we open screen, we couldn't find the submitter using search
> But if we unselect, submitter will be always visible using search
This one I managed to reproduce and indeed it's very weird... I will look into this right now 🕵️‍♀️

**Review by blazejkustra — APPROVED** [review]
LGTM!

File: src/libs/OptionsListUtils.ts
Comment (Julesssss) [reply src/libs/OptionsListUtils.ts:2205]: Yeah, correct. I'm not worried about this for now

**Comment by Julesssss** [comment]
Okay, this is looking good. Let's merge these changes to the feature branch so we can get the backend changes in place for you. Then we can handle the remaining bugs (recent empty, workspaces showing in list, ect)

**Review by Julesssss — APPROVED** [review]

**Pull request merged**
